[fs-extra] remove function should accept options

oganexon opened the pull request:
If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/rimraf/index.d.ts
https://github.com/jprichardson/node-fs-extra/blob/master/lib/remove/index.js

The `remove` function is a wrapper around `rimraf`.

Since `rimraf` accept parameters and `remove` is the universal version of it, it should also accept the same parameters.

Example of use:
```javascript
fs.remove(path, {
    glob: { dot: true }
  })
```

oganexon commented:
Update: now the `remove` function takes parameters with `fs-extra` type instead of `fs`.

oganexon commented:
I thought of another way, but could it work? I'm not very familiar with advanced types.
```typescript
export type RemoveOptions = rimraf.Options | {
    unlink?: typeof unlink;
    chmod?: typeof chmod;
    stat?: typeof stat;
    lstat?: typeof lstat;
    rmdir?: typeof rmdir;
    readdir?: typeof readdir;
}

```

peterblazejewicz commented on types/fs-extra/index.d.ts:
```diff
@@ -304,6 +307,27 @@ export interface ReadOptions {
     flag?: string;
 }
 
+export interface RemoveOptions {
```
Do we need all those options to mimick the rimraf details?
It happens that rimarf has dependency on the `glob`:
https://github.com/isaacs/rimraf/blob/master/package.json
but I think we should not just use same options, could be enough to just implement interface from the `options` actual usage here:
https://github.com/jprichardson/node-fs-extra/blob/a571007f7066eb49e93343cd7f36bc373bc8dfc2/lib/remove/rimraf.js#L27

oganexon replied:
I see.
But there's something I don't understand. Why is it that when I set a glob parameter, the function takes it into account?

oganexon replied:
I've removed the glob part.
Is it correct now?

peterblazejewicz replied:
yup, nearly there:
```ts
maxBusyTries?: number;
chmod?: typeof fs.chmod;
chmodSync?: typeof fs.chmodSync;
lstat?: typeof fs.lstat;
lstatSync?: typeof fs.lstatSync;
readdir?: typeof fs.readdir;
readdirSync?: typeof fs.readdirSync;
rmdir?: typeof fs.rmdir;
rmdirSync?: typeof fs.rmdirSync;
stat?: typeof fs.stat;
statSync?: typeof fs.statSync;
unlink?: typeof fs.unlink;
unlinkSync?: typeof fs.unlinkSync;
```
(to keep things together)

oganexon replied:
I propose not to put the fs typing but rather the fs-extra ones.
When using the library, it would require to have fs AND fs-extra since the the types are incompatibles and `typeof fs.[anything]` is pretty hard to reproduce because it has a lot of properties like `_promisify_`.
So we should use `rmdir` instead of `fs.rmdir` in my opinion.
However fs-extra does not have sync versions of the functions.

peterblazejewicz replied:
I think these should come from 'graceful-fs` (dependency):
https://github.com/jprichardson/node-fs-extra/blob/master/lib/remove/rimraf.js#L3
The import in the definition file is probably wrong.
What you'think?

oganexon replied:
That would be the ideal.
But I face a recurrent issue in my projects. Consider the following ts file:
```typescript
import * as gfs from 'graceful-fs'
import fs from 'fs-extra'

interface test {
  unlink: typeof gfs.unlink
}

const obj: test = {
  unlink: fs.unlink
}
```
An error is thrown on the `unlink` property of obj:
```
Property '__promisify__' is missing in type '{ (path: string | Buffer, callback: (err: ErrnoException) => void): void; (path: string | Buffer): Promise<void>; }' but required in type 'typeof unlink'.ts(2741)
fs.d.ts(726, 18): '__promisify__' is declared here.
test.ts(5, 3): The expected type comes from property 'unlink' which is declared here on type 'test'
```
The types of graceful-fs solve this issue whereas with graceful-fs or fs it's almost impossible to provide a clean function.

uniqueiniquity replied:
@oganexon @peterblazejewicz Is this ready to merge?

oganexon replied:
I think so but @peterblazejewicz doesn't seem to reply.
I've summarized the situation here:
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/43329#issuecomment-604549428

oganexon commented:
In order to solve the problem, we need to answer the following question:

Should we take the `graceful-fs` types or the `fs-extra` types ?

It seems logical to me to privilege the `fs-extra` types, being promises ready. `graceful-fs` functions still need the `__promisify__` property which isn't compatible at all with `fs-extra` functions.

To sum up, the aim is to avoid this kind of problem:
> But I face a recurrent issue in my projects. Consider the following ts file:
> 
> ```ts
> import * as gfs from 'graceful-fs'
> import fs from 'fs-extra'
> 
> interface test {
>   unlink: typeof gfs.unlink
> }
> 
> const obj: test = {
>   unlink: fs.unlink
> }
> ```
> 
> An error is thrown on the `unlink` property of obj:
> 
> ```
> Property '__promisify__' is missing in type '{ (path: string | Buffer, callback: (err: ErrnoException) => void): void; (path: string | Buffer): Promise<void>; }' but required in type 'typeof unlink'.ts(2741)
> fs.d.ts(726, 18): '__promisify__' is declared here.
> test.ts(5, 3): The expected type comes from property 'unlink' which is declared here on type 'test'
> ```
> 
> The types of fs-extra solve this issue whereas with graceful-fs or fs it's almost impossible to provide a clean function.

Why is my proposed implementation problematic?

```typescript
    unlink?: typeof unlink;
    chmod?: typeof chmod;
    stat?: typeof stat;
    lstat?: typeof lstat;
    rmdir?: typeof rmdir;
    readdir?: typeof readdir;
    unlinkSync?: typeof fs.unlinkSync;
    chmodSync?: typeof fs.chmodSync;
    statSync?: typeof fs.statSync;
    lstatSync?: typeof fs.lstatSync;
    rmdirSync?: typeof fs.rmdirSync;
    readdirSync?: typeof fs.readdirSync;
```

uniqueiniquity commented:
@oganexon @peterblazejewicz Here's my recommendation - unfortunately, the whole @types/fs-extra file is written with a suboptimal structure, which is getting in the way of solving this problem correctly.

Take `unlink` for example, as you have:
```ts
import fs from 'fs-extra'

fs.unlink()
```
Here, as it's currently written, I get two overloads for `unlink`, both from `fs-extra/index.d.ts`.
But the implementation from the non-`Promise`-returning overload comes from `fs` itself.
Therefore, I think it would better if, instead of overwriting the version that `fs` provides, this package simply adds on the `Promise`-returning overload.

It can do this by being structured as follows:
```ts
// this part stays mostly the same
/// <reference types="node" />
import * as fs from "graceful-fs";
export * as fs from "graceful-fs";

// this part changes
declare module "fs" {
  export function unlink(path: fs.PathLike): Promise<void>;
  // and others are done the same way
}
```
When doing it this way, calling `fs.unlink()` still gives two overloads - however, now one is from `fs-extra/index.d.ts` and one is from `node/fs.d.ts`.

When this is done, @peterblazejewicz's suggested implementation of `ReadOptions` works out correctly.

@peterblazejewicz, as a package owner, could you please coordinate the conversion of this module as I've suggested?

peterblazejewicz commented:
@uniqueiniquity thx for your comments and spending time on this, I was not exactly sure how to approach the problem with that definition file. Let's us try what you've suggested.
Thanks!

@oganexon 
Whan we have been suggested is to rewrite details to use module augmentation, son instead of having duplicate methods we should just add vai module augmentation ones that adds asynchronous execution via promise (so with added async/await out-of-the-box).

My only concern is about backward compatiblity with other code, as there are dozens of other packages listed as dependant on that little d.ts:
https://www.npmjs.com/package/@types/fs-extra

@oganexon can you please amend your changes according to @uniqueiniquity guideline. Then we will try to assert the impact, I could just look into other package usage of the code and we could amend tests to cover that other usage,

Thanks!

oganexon commented on types/fs-extra/index.d.ts:
```diff
@@ -113,151 +115,75 @@ export function pathExistsSync(path: string): boolean;
-
-export function open(path: string | Buffer, flags: string | number, callback: (err: NodeJS.ErrnoException, fd: number) => void): void;
-export function open(path: string | Buffer, flags: string | number, mode: number, callback: (err: NodeJS.ErrnoException, fd: number) => void): void;
+export function mkdir(path: string | Buffer, mode?: number | string): Promise<void>;
```
The `mode` option didn't appear in the promise version, I added it. I think that was an oversight.

oganexon commented on types/fs-extra/index.d.ts:
```diff
@@ -304,6 +230,22 @@ export interface ReadOptions {
 
+export interface RemoveOptions {
+    maxBusyTries?: number;
+    unlink?: typeof fs.unlink | typeof unlink;
```
My implementation may not be the best. The functions have to be the union of the callback and promise types.

oganexon commented:
I did what uniqueiniquity said.
I deleted all the callback versions and modified the arguments by taking those of the `fs` module.
But it's beyond my scope, I can't solve all the problems, I need your help @peterblazejewicz.

oganexon commented:
@peterblazejewicz are you still working on the problem?  Or will it never be solved?
